### PR TITLE
docs: document migrating from PolymerTemplate

### LINKED ISSUE #1190: Document migrating from TemplateRenderer to LitRenderer
`TemplateRenderer` is deprecated in favour of `LitRenderer`.
Add short migration instructions for what needs to be changed:

Before:
```java
TemplateRenderer.of("<order-card"
	+ "  header='[[item.header]]'"
	+ "  order-card='[[item.orderCard]]'"
	+ "  on-card-click='cardClick'>"
	+ "</order-card>");
```

After:
```java
LitRenderer.of("<order-card"
	+ "  .header='${item.header}'"
	+ "  .orderCard='${item.orderCard}'"
	+ "  @card-click='${cardClick}'>"
	+ "</order-card>");
```

### LINKED ISSUE #1204: Document migrating from PolymerTemplate to LitTemplate
`PolymerTemplate` is deprecated in favor of `LitTemplate`.
Add migration instructions for what needs to be changed.

Make sure to cover `TemplateModel`, `@Include`, `@Encode`

### PR DESCRIPTION
**Pull request by tomivirkki**
Closes #1204

### CODE REVIEW

File: articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc [hunk 1]
```diff
@@ -0,0 +1,164 @@
+In the latest Vaadin versions the <<{articles}/flow/templates/polymer#, [classname]`PolymerTemplate`>> class is deprecated and using a <<{articles}/flow/templates#, [classname]`LitTemplate`>> is recommended instead.
+
+In trivial cases, it is enough to replace the [classname]`PolymerTemplate` superclass with [classname]`LitTemplate` and then to migrate the `PolymerElement` based client view to base on `LitElement` instead.
```
Comment (dougjewsbury) [inline articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc:3]:
```suggestion
In trivial cases, it is enough to replace the [classname]`PolymerTemplate` superclass with [classname]`LitTemplate` and then to migrate the `PolymerElement` based client view to be based on `LitElement` instead.
```

File: articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+html`<button @click="${e => this.$server.handleClick()}">Click me</button>`;
+----
+
+[discrete]
+=== Migrating From TemplateModel
+
+[classname]`LitTemplate` does not support the [classname]`TemplateModel` API and its utilities.
+With Vaadin 23, it is recommended to handle server-to-client data communication with the Vaadin <<{articles}/flow/element-api/properties-attributes#, Element APIs>> instead.
+
+For example, the [methodname]`getElement().setProperty("propertyName", value)` method can be used for setting primitive property values for the client side view.
```
Comment (dougjewsbury) [inline articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc:71]:
```suggestion
For example, the [methodname]`getElement().setProperty("propertyName", value)` method can be used for setting primitive property values for the client-side view.
```

File: articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    JsonObject json = Json.createObject();
+    json.addProperty("id", order.getId());
+    json.addProperty("name", order.getName());
+    json.addProperty("totalPrice", new CurrencyFormatter().encode(order.getTotalPrice()));
+    return json;
+}
+----
+
+If a property value gets modified by the client logic, the update needs to be communicated back to the server.
+For this, you can use a <<{articles}/flow/element-api/client-server-rpc#clientcallable-annotation, [classname]`@ClientCallable`>> method.
```
Comment (dougjewsbury) [inline articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc:119]:
```suggestion
For this, you can use a <<{articles}/flow/element-api/client-server-rpc#clientcallable-annotation, [annotationname]`@ClientCallable`>> method.
```

**Review by joheriks — CHANGES_REQUESTED** [review]

File: articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+            <div>
+            ...`
+
+/* After */
+class HelloWorld extends LitElement {
+    render() {
+        return html`
+            <div>
+            ...`
+----
```
[the remark this answers is not in the document]
Comment (tomivirkki) [reply articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc:34]: Done

File: articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+private JsonObject toJson(Order order) {
+    JsonObject json = Json.createObject();
+    json.addProperty("id", order.getId());
+    json.addProperty("name", order.getName());
+    json.addProperty("totalPrice", new CurrencyFormatter().encode(order.getTotalPrice()));
+    return json;
+}
+----
+
```
[the remark this answers is not in the document]
Comment (tomivirkki) [reply articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc:125]: Not sure if we want to recommend two different ways of doing this? @Artur- ?

[the remark this answers is not in the document]
Comment (joheriks) [reply articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc:125]: I was thinking of it more as a hint than a recommended way. For example, a project may already have some DTOs that could be reused for this purpose (as was the case for Taskmob).

File: articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc [hunk 6]
```diff
@@ -15,20 +15,28 @@ public class HelloWorld extends LitTemplate {
...
 class HelloWorld extends LitElement {
     render() {
         return html`
             <div>
             ...`
 ----
 
+.Update @Id package name
+[NOTE]
+If you use `@Id` for <<{articles}/flow/templates/components#, binding to template components>>, note that the package name should be changed from `com.vaadin.flow.component.polymertemplate.Id` to `com.vaadin.flow.component.template.Id` for [classname]`LitTemplate`.
```
Comment (jouni) [inline articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc:38]: Feel free to use the recently added `[annotationname]` style when appropriate.

```suggestion
If you use [annotationname]`@Id` for <<{articles}/flow/templates/components#, binding to template components>>, note that the package name should be changed from `com.vaadin.flow.component.polymertemplate.Id` to `com.vaadin.flow.component.template.Id` for [classname]`LitTemplate`.
```

File: articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (tomivirkki) [reply articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc:125]: Ah, got it. Added a note about using DTO.

File: articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+As an alternative to building a [classname]`JsonObject` manually, you can create a Java DTO class containing only the fields needed by the client and passing instances of that DTO to [methodname]`setPropertyBean`.
+
+If a property value gets modified by the client logic, the update needs to be communicated back to the server.
+For this, you can use a <<{articles}/flow/element-api/client-server-rpc#clientcallable-annotation, [annotationname]`@ClientCallable`>> method.
+
+[source,java]
+----
+@ClientCallable
+public void setProperty(String value) {
+    getElement().setProperty("propertyName", value);
```
Comment (joheriks) [inline articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc:137]: Would this actually be a typical case? If the property was updated on the client's initiative, I would think the server wants to do something else than just setting it again on the client. Perhaps something like 
```java
@ClientCallable
public void setProperty(String value) {
    // handle updated property on the server
}
```
so it doesn't look like something one "should do".

Comment (tomivirkki) [reply articles/v14-23-upgrade/final-steps/instructions/_polymertemplate.adoc]: Done

**Review by joheriks — APPROVED** [review]

**Pull request merged**